private/pkg/git/tree.go

```diff
@@ -0,0 +1,110 @@
+// Copyright 2020-2023 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package git
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+)
+
+type tree struct {
+	hash  Hash
+	nodes []Node
+}
+
+func parseTree(hash Hash, data []byte) (*tree, error) {
+	t := &tree{
+		hash: hash,
+	}
+	/*
+		data is in the format
+			<mode><space><name>\0<hash>
+		repeated
+	*/
+	for len(data) > 0 {
+		// We can find the \0 character before the <hash>
+		// and slice to the index of \0 + the length of a hash.
+		// That gives us a single node.
+		i := bytes.Index(data, []byte{0})
+		if i == -1 {
+			return nil, errors.New("malformed tree")
+		}
+		length := i + 1 + hashLength
+		node, err := parseTreeNode(data[:length])
+		if err != nil {
+			return nil, fmt.Errorf("malformed tree: %w", err)
+		}
+		t.nodes = append(t.nodes, node)
+		data = data[length:]
+	}
+	return t, nil
+}
+
+func (t *tree) Hash() Hash {
+	return t.hash
+}
+
+func (t *tree) Nodes() []Node {
+	return t.nodes
+}
+
+func (t *tree) Traverse(objectReader ObjectReader, names ...string) (Node, error) {
+	return traverse(objectReader, t, names...)
+}
+
+func traverse(
+	objectReader ObjectReader,
+	root Tree,
+	names ...string,
+) (Node, error) {
+	// split by the name of the next node we're looking for
+	// and the names of the descendant nodes
+	name := names[0]
+	if len(names) >= 2 {
+		names = names[1:]
+	} else {
+		names = nil
+	}
```

Comment:
I know I suggested receiving the whole path, but these renames still apply.
```suggestion
func traverse(
	objectReader ObjectReader,
	root Tree,
	pathComponents ...string,
) (Node, error) {
	if len(pathComponents) == 0 {
		return nil, errors.New("empty path components") // avoid panic
	}
	// split by the name of the next node we're looking for
	// and the names of the descendant nodes
	targetNode := pathComponents[0]
	var nextTargetNodes []string
	if len(pathComponents) >= 2 {
		nextTargetNodes := pathComponents[1:]
	}
```